site/content/contribute/more-info/server/tests.md

```diff
@@ -21,10 +21,79 @@ A flaky test is one that exhibits both passing and failing results when run mult
    - Copy the test failure message into the JIRA ticket description
    - Add the `flaky-test` and `triage-global` labels
    - Create a PR to skip the test by adding:
+
      ```go
      t.Skip("https://mattermost.atlassian.net/browse/MM-XXXXX")
      ```
+
      where MM-XXXXX is your JIRA ticket number
    - Link the JIRA ticket in the skip message for tracking
 
 This process helps us track and systematically address flaky tests while preventing them from blocking development work.
+
+## Writing Parallel Tests
+
+Leveraging parallel tests can drastically reduce execution time for entire test packages, such as `api4` and `app`, which are notably heavy with hundreds of tests. However, careful implementation is essential to ensure reliability and prevent flakiness. Follow these guidelines when writing parallel tests:
+
+### Enabling Parallel Tests
+
+Ensure parallel tests run behind an explicit environment flag to control execution:
+
+```go
+// In api4 and app packages, check mainHelper.Options.RunParallel
+if mainHelper.Options.RunParallel {
+    t.Parallel()
+}
+
+// In sqlstore package, a global enableFullyParallelTests variable is used instead
+if enableFullyParallelTests {
+    t.Parallel()
+}
```

Comment:
Is this something each test has to do for itself? Could we add a `Parallel` method to the commonly used `testHelper` struct that encodes this logic instead?

Reply:
I'm afraid so, as I don't think that adding it to the helper would be a safe default because whether parallelism can be used depends on the specific test implementation.

Reply:
No, what I mean is can we expose this:
```go
func (th *testHelper) Parallel(t) {
    if mainHelper.Options.RunParallel {
        t.Parallel()
    }
}
```

and just do this from the test?
```go
func TestBlah(t *testing.T) {
    th := setupHelper()
    th.Parallel(t)
    // ...
}
```

Reply:
@lieut-data For simplicity (there are hundreds of files to modify as a result of this), I propose adding the helper function to `mainHelper` as well, so we can do:

```go
func TestExample(t *testing.T) {
  mainHelper.Parallel(t)
}
```

Alternatively, we can also do what you suggested:

```go
func TestExample(t *testing.T) {
  th := Setup(t)
  th.Parallel(t)
}
```

But it just becomes a pain to update automatically now since pattern matching becomes fairly complex.

Reply:
I pushed these changes at https://github.com/mattermost/mattermost/pull/30078